### PR TITLE
Add builtins docs

### PR DESCRIPTION
**Pull request by zbraniecki**
First pass, @Pike , @stasm ?

### CODE REVIEW

File: docs/builtins.rst [hunk 1]
```diff
@@ -0,0 +1,217 @@
+================
+Builtins in L20n
```
Comment (stasm) [inline docs/builtins.rst]: Let's use the word `Function` to describe the general class of these things, and then `Built-in Function` to describe those that are provided by `MessageContext` by default.

File: docs/builtins.rst [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+================
+
+L20n ships with a number of built-in functions that can be used from within
+localization messages.
```
Comment (stasm) [inline docs/builtins.rst]: I'm missing a sentence about the purpose of `Functions`.  Something about formatting data according to the current language's rules would be enough.

File: docs/builtins.rst [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+
+The list of built-ins is extensible and environments may want to add additional
```
Comment (stasm) [inline docs/builtins.rst]: `The list of built-in functions is…`

File: docs/builtins.rst [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+
+Default built-ins
+=================
+
+Default built-ins are very generic and should be applicable to any translation
+environment.
+
+``NUMBER``
+    Formats a number to a string in a given locale.
+    Accepts the same parameters as `Intl.NumberFormat`_ API.
```
Comment (stasm) [inline docs/builtins.rst]: I know it's repetition but I think we should provide a quick summary of the available parameters in form of a table for each function.

File: docs/builtins.rst [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+built-ins designated to aid localizers writing translations targeted
+for a given environment.
+
+
+How to use Built-ins
+====================
+
+Localizers can reference builtin in a message within a placeable and either
+display the return value, or use it in the selector expression to select
+the right value.
```
Comment (stasm) [inline docs/builtins.rst]: `FTL Functions can only be called inside of placeables. Use them to return a value to be interpolated in the message or as selectors in select expressions.  Functions may (but don't have to) accept positional and keyword arguments. See the reference below for more information about the arguments accepted for each built-in function.`

File: docs/builtins.rst [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+
+      users = { LIST($users) }
+
+      users2 = { $users }
+
+
+Partially resolved built-ins
+============================
+
+In the future we'll want to allow for partially resolved builtins to be
```
Comment (stasm) [inline docs/builtins.rst]: That's already possible right now! https://github.com/l20n/l20n.js/blob/3ba0b9ad8f20a9540eb7ef6c4bda4604af4677e1/src/intl/polyfill.js#L6-L7

File: docs/builtins.rst [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+constructed by the developer and passed to the localization.
+This scenario will be used when the developer wants to define the
+default parameters for the built-in, but allow the localizer
+to override them if needed.
+
+It may look like this::
+
+  main.js:
+
+  let s = document.l10n.formatValue('key1', {
```
Comment (stasm) [inline docs/builtins.rst]: Let's keep this doc scoped to FTL and the `MessageContext` class.  Change this exampe to use `ctx.format()`?

File: docs/builtins.rst [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+  key1 = Current date is { $date }
+
+  # If the localizer needs to change the way we display month to fit the text
+  key1 = Current date is { DATETIME($date, month=short) }
+
+
+Gecko runtime specific built-ins
+================================
+
+At the moment Gecko supports the following OS-specific built-ins:
```
Comment (stasm) [inline docs/builtins.rst]: Should this go into `gecko.rst`?

Comment (stasm) [inline docs/builtins.rst]: Also, I'm thinking we should rename this function to `PLATFORM` and return full names: `windows`, 'macos`, 'gnome', 'unix`.

File: docs/builtins.rst [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+====================
+
+Localizers can reference builtin in a message within a placeable and either
+display the return value, or use it in the selector expression to select
+the right value.
+
+Examples::
+
+
+  today-is = Today is { DATETIME($date, weekday=long) }
```
Comment (stasm) [inline docs/builtins.rst]: Our current syntax requires `:` for the keyword arguments and quotes:

    today-is = Today is { DATETIME($date, weekday: "long") } 

https://github.com/l20n/l20n.js/blob/3ba0b9ad8f20a9540eb7ef6c4bda4604af4677e1/src/ftl/ast/parser.js#L503

We should probably change the EBNF grammar: https://github.com/l20n/spec/blob/126b422ae146cdb81eed38484eca8d732cec46a6/grammar.ebnf#L49

File: docs/builtins.rst (hunk 8 above)
Comment (zbraniecki) [inline docs/builtins.rst]: Not sure. @Pike ?

Comment (stasm) [inline docs/builtins.rst]: I filed https://bugzilla.mozilla.org/show_bug.cgi?id=1311666 to track the renaming.  Let's not block on it.

File: docs/builtins.rst [hunk 10]
```diff
@@ -0,0 +1,296 @@
...
+
+      ''currencyDisplay''
+      ''useGrouping''
+      ''minimumIntegerDigits''
+      ''minimumFractionDigits''
+      ''maximumFractionDigits''
+      ''minimumSignificantDigits''
+      ''maximumSignificantDigits''
+
+    Devloper parameters::
```
Comment (stasm) [inline docs/builtins.rst]: `Developer` (typo here and elsewhere).

File: docs/builtins.rst [hunk 11]
```diff
@@ -0,0 +1,296 @@
...
+
+
+  today-is = Today is { DATETIME($date) }
+
+
+Function parameters
+===================
+
+Functions may (but don't have to) accept positional and keyword arguments.
+Some parameters are only available to developers passing the Function as
```
Comment (stasm) [inline docs/builtins.rst]: Maybe: `Some keyword parameters are only available to developers and cannot be used inside of FTL translations.  Instead they can be set on external arguments passed into translations by the developer.`

File: docs/builtins.rst [hunk 12]
```diff
@@ -0,0 +1,296 @@
...
+
+    Devloper parameters::
+
+      ''type''
+
+    See the `Intl.PluralRules`_ for the description of the parameters.
+
+``LIST``
+    Formats a list to a string in a given locale.
+    Accepts the same parameters as `Intl.ListFormat`_ API.
```
Comment (stasm) [inline docs/builtins.rst]: Remove as line 139 has the same info.

File: docs/builtins.rst [hunk 13]
```diff
@@ -0,0 +1,296 @@
...
+``LEN``
+    Returns the number that represents the length of the list argument.
+    Similar to JS `Array.length`_ API.
+
+    Example::
+
+      unread-emails = Number of unread emails: { LEN($emails) }
+
+``TAKE``
+    Returns a slice of a list.
```
Comment (stasm) [inline docs/builtins.rst]: `... starting from the first element`.  Please describe the parameters that `TAKE` and `DROP` accept.

File: docs/builtins.rst [hunk 14]
```diff
@@ -0,0 +1,304 @@
...
+  key1 = Today is { DATETIME($day, weekday: "short") }
+
+
+
+Gecko runtime specific functions
+================================
+
+At the moment Gecko runtime adds the following functions:
+
+``PLATFORM``
```
Comment (stasm) [inline docs/builtins.rst]: The bug about the renaming hasn't yet been resolved, so it's up to you if you want to use the new name here or the old one.

**Pull request merged**